tests/components/tasmota/test_common.py

```diff
@@ -0,0 +1,328 @@
+    unique_id = f"{config[CONF_MAC]}_{domain}_0"
+
+    data = json.dumps(config)
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config", data)
+    await hass.async_block_till_done()
+
+    device = device_reg.async_get_device(set(), {("mac", config[CONF_MAC])})
+    assert device is not None
+    assert entity_reg.async_get_entity_id(domain, "tasmota", unique_id)
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config", "")
+    await hass.async_block_till_done()
+
+    device = device_reg.async_get_device(set(), {("mac", config[CONF_MAC])})
+    assert device is None
+    assert not entity_reg.async_get_entity_id(domain, "tasmota", unique_id)
+
+
+async def help_test_entity_id_update_subscriptions(
+    hass, mqtt_mock, domain, config, topics=None
+):
+    """Test MQTT subscriptions are managed when entity_id is updated."""
+    entity_reg = await hass.helpers.entity_registry.async_get_registry()
+
+    config = copy.deepcopy(config)
+    data = json.dumps(config)
+
+    mqtt_mock.async_subscribe.reset_mock()
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config", data)
+    await hass.async_block_till_done()
+
+    topics = [get_topic_tele_state(config), get_topic_tele_will(config)]
+    assert len(topics) > 0
+
+    state = hass.states.get(f"{domain}.test")
+    assert state is not None
+    assert mqtt_mock.async_subscribe.call_count == len(topics)
+    for topic in topics:
+        mqtt_mock.async_subscribe.assert_any_call(topic, ANY, ANY, ANY)
+    mqtt_mock.async_subscribe.reset_mock()
+
+    entity_reg.async_update_entity(f"{domain}.test", new_entity_id=f"{domain}.milk")
+    await hass.async_block_till_done()
+
+    state = hass.states.get(f"{domain}.test")
+    assert state is None
+
+    state = hass.states.get(f"{domain}.milk")
+    assert state is not None
+    for topic in topics:
+        mqtt_mock.async_subscribe.assert_any_call(topic, ANY, ANY, ANY)
+
+
+async def help_test_entity_id_update_discovery_update(hass, mqtt_mock, domain, config):
+    """Test MQTT discovery update after entity_id is updated."""
+    entity_reg = await hass.helpers.entity_registry.async_get_registry()
+
+    config = copy.deepcopy(config)
+    data = json.dumps(config)
+
+    topic = get_topic_tele_will(config)
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config", data)
+    await hass.async_block_till_done()
+
+    async_fire_mqtt_message(hass, topic, get_state_online(config))
+    state = hass.states.get(f"{domain}.test")
+    assert state.state != STATE_UNAVAILABLE
+
+    async_fire_mqtt_message(hass, topic, get_state_offline(config))
+    state = hass.states.get(f"{domain}.test")
+    assert state.state == STATE_UNAVAILABLE
+
+    entity_reg.async_update_entity(f"{domain}.test", new_entity_id=f"{domain}.milk")
+    await hass.async_block_till_done()
+
+    assert config[CONF_PREFIX][PREFIX_TELE] != "tele2"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)